pkg/declarative/container.go

```diff
@@ -0,0 +1,226 @@
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/image"
+	"github.com/docker/docker/client"
+	"gopkg.in/yaml.v2"
+)
+
+type Containerrunner struct {
+	ContainerId string
+	Image       string
+}
+
+func (r *Containerrunner) Setup(beforeScript string) error {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return fmt.Errorf("error creating Docker Client: %v", err)
+	}
+
+	// Pull the image
+	pullRes, err := cli.ImagePull(context.Background(), r.Image, image.PullOptions{})
+	if err != nil {
+		return fmt.Errorf("error pulling Docker Image: %v", err)
+	}
+
+	// Read the response of body inorder to download the image
+	defer pullRes.Close()
+	_, err = io.Copy(io.Discard, pullRes)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Create the container with name "alpine-container-by-event-generator"
+	resp, err := cli.ContainerCreate(context.Background(),
+		&container.Config{
+			Image: r.Image,
+			Cmd:   []string{"sleep", "1h"},
+		}, nil, nil, nil, "alpine-container-by-event-generator")
+
+	if err != nil {
+		return fmt.Errorf("error creating Docker Container: %v", err)
+	}
+
+	// Store created container id
+	r.ContainerId = resp.ID
+
+	// Path of the event-generator executable
+	path_eg, err := os.Executable()
+	if err != nil {
+		return fmt.Errorf("error extracting path of event-generator executable")
+	}
+
+	// Copy the event-generator executable into the container
+	err = exec.Command("docker", "cp", path_eg, fmt.Sprintf("%s:%s", r.ContainerId, "/")).Run()
+	if err != nil {
+		return fmt.Errorf("error copying file to container: %v", err)
+	}
+
+	// Start the container
+	err = cli.ContainerStart(context.Background(), r.ContainerId, container.StartOptions{})
+	if err != nil {
+		return fmt.Errorf("error starting Docker container: %v", err)
+	}
+
+	// Run beforescript
+	cmd := []string{"sh", "-c", beforeScript}
+
+	execConfig := container.ExecOptions{
+		Cmd: cmd,
+	}
+
+	execID, err := cli.ContainerExecCreate(context.Background(), r.ContainerId, execConfig)
+	if err != nil {
+		return fmt.Errorf("error creating Docker exec: %v", err)
+	}
+
+	err = cli.ContainerExecStart(context.Background(), execID.ID, container.ExecStartOptions{})
+	if err != nil {
+		return fmt.Errorf("error starting Docker exec: %v", err)
+	}
+
+	// Use this channel to wait for the exec instance to complete i.e, beforescript
+	done := make(chan error, 1)
+	go func() {
+		for {
+			inspectResp, err := cli.ContainerExecInspect(context.Background(), execID.ID)
+			if err != nil {
+				done <- fmt.Errorf("error inspecting Docker exec: %v", err)
+				return
+			}
+
+			if !inspectResp.Running {
+				if inspectResp.ExitCode != 0 {
+					done <- fmt.Errorf("before execution failed with exit code %v", inspectResp.ExitCode)
+				} else {
+					done <- nil
+				}
+				return
+			}
+		}
+	}()
+
+	err = <-done
```

Comment:
In the `ExecuteStep` we run the rule in the container with the host runner. I will delegate the `setup` and `cleanup` steps to it.